internal/wal_show_handler.go

```diff
@@ -0,0 +1,212 @@
+		SegmentRangeSize: uint64(walSegments.maxSegmentNo-walSegments.minSegmentNo) + 1,
+		Status:           TimelineOkStatus,
+	}
+
+	missingSegments, err := walSegments.FindMissingSegments()
+	if err != nil {
+		return nil, err
+	}
+	timelineInfo.MissingSegments = make([]string, 0, len(missingSegments))
+	for _, segment := range missingSegments {
+		timelineInfo.MissingSegments = append(timelineInfo.MissingSegments, segment.GetFileName())
+	}
+
+	if len(timelineInfo.MissingSegments) > 0 {
+		timelineInfo.Status = TimelineLostSegmentStatus
+	}
+
+	// set parent timeline id and timeline switch LSN if have .history record available
+	if len(historyRecords) > 0 {
+		switchHistoryRecord := historyRecords[len(historyRecords)-1]
+		timelineInfo.ParentId = switchHistoryRecord.timeline
+		timelineInfo.SwitchPointLsn = switchHistoryRecord.lsn
+	}
+	return timelineInfo, nil
+}
+
+// WalSegmentsSequence represents some collection of wal segments with the same timeline
+type WalSegmentsSequence struct {
+	timelineId        uint32
+	walSegmentNumbers map[WalSegmentNo]bool
+	minSegmentNo      WalSegmentNo
+	maxSegmentNo      WalSegmentNo
+}
+
+func NewSegmentsSequence(id uint32, segmentNo WalSegmentNo) *WalSegmentsSequence {
+	walSegmentNumbers := make(map[WalSegmentNo]bool)
+	walSegmentNumbers[segmentNo] = true
+
+	return &WalSegmentsSequence{
+		timelineId:        id,
+		walSegmentNumbers: walSegmentNumbers,
+		minSegmentNo:      segmentNo,
+		maxSegmentNo:      segmentNo,
+	}
+}
+
+// AddWalSegmentNo adds the provided segment number to collection
+func (data *WalSegmentsSequence) AddWalSegmentNo(number WalSegmentNo) {
+	data.walSegmentNumbers[number] = true
+	if data.minSegmentNo > number {
+		data.minSegmentNo = number
+	}
+	if data.maxSegmentNo < number {
+		data.maxSegmentNo = number
+	}
+}
+
+// FindMissingSegments finds missing segments in range [minSegmentNo, maxSegmentNo]
+func (data *WalSegmentsSequence) FindMissingSegments() ([]*WalSegmentDescription, error) {
+	startWalSegment := WalSegmentDescription{number: data.maxSegmentNo, timeline: data.timelineId}
+
+	walSegments := make(map[WalSegmentDescription]bool, len(data.walSegmentNumbers))
+	for number := range data.walSegmentNumbers {
+		segment := WalSegmentDescription{number: number, timeline: data.timelineId}
+		walSegments[segment] = true
+	}
+
+	// create WAL segment runner to run on single timeline
+	walSegmentRunner := NewWalSegmentRunner(startWalSegment, walSegments, data.minSegmentNo)
+	missingSegments := make([]*WalSegmentDescription, 0)
+	for {
+		if _, err := walSegmentRunner.MoveNext(); err != nil {
+			switch err := err.(type) {
+			case WalSegmentNotFoundError:
+				// force switch to the next WAL segment
+				walSegmentRunner.ForceMoveNext()
+				missingSegments = append(missingSegments, &walSegmentRunner.currentWalSegment)
+			case ReachedStopSegmentError:
+				// Can't continue because reached stop segment, stop at this point
+				return missingSegments, nil
+			default:
```

Comment:
If MoveNext return nil, it will go to default and finish the loop.

Reply:
It meant to scan until we reach stop segment